metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala

```diff
@@ -924,4 +920,84 @@ object DebugProvider {
       extends Exception(
         "Build misconfiguration. No semanticdb can be found for you file, please check the doctor."
       )
+
+  sealed abstract class ClassSearch[A](debugProvider: DebugProvider)(implicit
+      ec: ExecutionContext
+  ) {
+    protected def f(): Try[A]
+    protected def dapSessionParams(res: A): Future[DebugSessionParams]
+    private def maybeCreateDapSession(
+        res: Future[Option[A]]
+    ): Future[Option[DebugSession]] = {
+      res.flatMap {
+        case Some(args) =>
+          dapSessionParams(args)
+            .flatMap(debugProvider.asSession(_))
+            .map(Some(_))
+        case None => Future.successful(None)
+      }
+    }
+    private lazy val result: Future[Try[A]] = Future(f())
+    private lazy val retryResult: Future[Option[A]] =
+      result.flatMap(debugProvider.retryAfterRebuild(_, f).map(_.toOption))
+    def dapSession: Future[Option[DebugSession]] = maybeCreateDapSession(
+      result.map(_.toOption)
+    )
+    def retryDapSession: Future[Option[DebugSession]] = maybeCreateDapSession(
+      retryResult
+    )
+  }
+
+  final class MainClassSearch(
+      debugProvider: DebugProvider,
+      params: DebugUnresolvedMainClassParams,
+  )(implicit ec: ExecutionContext)
+      extends ClassSearch[List[(ScalaMainClass, b.BuildTarget)]](
+        debugProvider
+      ) {
+    override protected def f(): Try[List[(ScalaMainClass, b.BuildTarget)]] =
+      debugProvider.findMainClassAndItsBuildTarget(params)
+    override protected def dapSessionParams(
+        res: List[(ScalaMainClass, b.BuildTarget)]
+    ): Future[DebugSessionParams] =
+      debugProvider.buildMainClassParams(res, params)
+  }
+
+  final class TestClassSearch(
+      debugProvider: DebugProvider,
+      params: DebugUnresolvedTestClassParams,
+  )(implicit ec: ExecutionContext)
+      extends ClassSearch[List[(String, b.BuildTarget)]](debugProvider) {
+    override protected def f(): Try[List[(String, b.BuildTarget)]] =
+      debugProvider.findTestClassAndItsBuildTarget(params)
+    override protected def dapSessionParams(
+        res: List[(String, b.BuildTarget)]
+    ): Future[DebugSessionParams] =
+      debugProvider.buildTestClassParams(res, params)
+  }
+
+  def getResultFromSearches[A](searches: List[ClassSearch[A]])(
+      default: => Future[DebugSession]
+  )(implicit ec: ExecutionContext): Future[DebugSession] = {
+    Future
+      .sequence(searches.map(_.dapSession))
```

Comment:
Potentially this could start multiple sessions (for example if we have duplicate project like two metals projects in the same workspace). Ideally, we would only invoke `_.dapSession` only if the previous one was not successful in finding the class to run.